### PR TITLE
TTS | Validation mode base

### PR DESCRIPTION
**Pull request by sudiksha27**
## General
**Before this PR**:
There was not a way to turn on transactions 4 without verification.

**After this PR**:
Added hard-coded verification mode.
I do not want to make this configurable just yet while transactions 4 is not actually ready.

**Priority**:
Before end of this week would have been ideal.

**Concerns / possible downsides (what feedback would you like?)**:
Double check the handling of unknown status

**Is documentation needed?**:
I do not think so.

## Compatibility
~**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:~

~**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:~

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
When the config changes, there will be transactions that have verification on co-existing with transactions that do not have verification.
Seeing this is already best-effort, it should not be a problem.

~**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:~

**Does this PR need a schema migration?**
No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
Pending

**What was existing testing like? What have you done to improve it?**:
Adding new tests

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
Because we first query the _txn2 table and then {_kct, _kat}, we can have more information but never less.

~**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:~

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
We do have a metric for failure, it might be too noisy to log/report on successful verifications. Not opposed to doing that initially though.

~**Has the safety of all log arguments been decided correctly?**:~

**Will this change significantly affect our spending on metrics or logs?**:
Adding one series

**How would I tell that this PR does not work in production? (monitors, etc.)**:
Another metric for success might be helpful here. I shall add that now.

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
Rolling back is fine. This one actually does not change anything in prod right now.

~**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:~

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
I am a bit skeptical how this could impact the perf but it caching in {_kct, _kat} work as expected the impact should not be noticeable.

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
Potentially. 

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
We would have to rollback from txn 4

## Development Process
**Where should we start reviewing?**:
`KnowledgeableTimestampExtractingAtomicTable`

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java [hunk 1]
```diff
@@ -22,24 +22,38 @@
...
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
 import java.util.Comparator;
 import java.util.Map;
+import java.util.Optional;
 import java.util.stream.StreamSupport;
 
 public class KnowledgeableTimestampExtractingAtomicTable implements AtomicTable<Long, Long> {
+    private static final SafeLogger log = SafeLoggerFactory.get(KnowledgeableTimestampExtractingAtomicTable.class);
+    private static final boolean IS_VALIDATION_MODE = true;
+
```
Comment (sudiksha27) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java:45]: Not super keen on making this configurable just yet

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java [hunk 2]
```diff
@@ -77,11 +97,59 @@ public ListenableFuture<Long> get(Long startTimestamp) {
...
+                    .collectToMap();
+            return AtlasFutures.allAsMap(futures, MoreExecutors.directExecutor());
+        }
+    }
+
+    private Optional<Long> verifyAndGetCommitTs(long startTimestamp, TransactionStatus transactionStatus) {
+        Optional<Long> maybeGetCommitTs = Optional.ofNullable(TransactionStatusUtils.getCommitTsFromStatus(
+                startTimestamp, transactionStatus, knownAbandonedTransactions::isKnownAbandoned));
+
+        maybeGetCommitTs.ifPresent(commitTs -> {
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java:122]: maybe also check if it's still in progress?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java [hunk 3]
```diff
@@ -77,11 +97,59 @@ public ListenableFuture<Long> get(Long startTimestamp) {
...
+            boolean isConcluded = knownConcludedTransactions.isKnownConcluded(
+                    startTimestamp, KnownConcludedTransactions.Consistency.REMOTE_READ);
+
+            if (!isConcluded) {
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java:126]: We can also check even if not concluded that we don't get abandoned && committed

**Review by gmaretic — APPROVED** [review]
Just nits

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java (hunk 3 above)
Comment (sudiksha27) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java:126]: The thing is we cannot ask abandoned cache for status of a non concluded timestamp. So we will have to talk to the store and that would carry perf penalty.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java [hunk 4]
```diff
@@ -77,11 +97,59 @@ public ListenableFuture<Long> get(Long startTimestamp) {
...
+        if (IS_VALIDATION_MODE) {
+            // No verification for in-progress transactions right now.
+            return Futures.transform(
+                    delegate.get(keys),
+                    statuses -> KeyedStream.stream(statuses)
+                            .map(this::verifyAndGetCommitTs)
+                            .flatMap(Optional::stream)
+                            .collectToMap(),
+                    MoreExecutors.directExecutor());
+        } else {
```
Comment (Sam-Kramer) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java:109]: super style nit, but you do not need the `else` !

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java (hunk 1 above)
Comment (Sam-Kramer) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/KnowledgeableTimestampExtractingAtomicTable.java:45]: It's probably worth doing this IMO just b/c it gives us an escape hatch in a p0 scenario in-case the verification really slams the KVS.

**Review by Sam-Kramer — APPROVED** [review]

**Pull request merged**